packages/gatsby-remark-images-contentful/src/index.js

```diff
@@ -0,0 +1,186 @@
+const select = require(`unist-util-select`)
+const sharp = require(`sharp`)
+const axios = require(`axios`)
+const _ = require(`lodash`)
+const Promise = require(`bluebird`)
+const cheerio = require(`cheerio`)
+const { buildResponsiveSizes } = require(`./utils`)
+
+// If the image is hosted on contentful
+// 1. Find the image file
+// 2. Find the image's size
+// 3. Filter out any responsive image sizes that are greater than the image's width
+// 4. Create the responsive images.
+// 5. Set the html w/ aspect ratio helper.
+
+module.exports = async (
+  { files, markdownNode, markdownAST, pathPrefix, getNode, reporter },
+  pluginOptions
+) => {
+  const defaults = {
+    maxWidth: 650,
+    wrapperStyle: ``,
+    backgroundColor: `white`,
+    linkImagesToOriginal: true,
+    showCaptions: false,
+    pathPrefix,
+  }
+
+  const options = _.defaults(pluginOptions, defaults)
+
+  // This will only work for markdown syntax image tags
+  const markdownImageNodes = select(markdownAST, `image`)
+
+  // This will also allow the use of html image tags
+  const rawHtmlNodes = select(markdownAST, `html`)
+
+  const generateImagesAndUpdateNode = async function(node, resolve) {
+    // Ingonre if it is not contentful image
+    if (node.url.indexOf(`images.ctfassets.net`) === -1) {
+      return resolve()
+    }
+
+    const metaReader = sharp()
+
+    const response = await axios({
+      method: `GET`,
+      url: `https:${node.url}`, // for some reason there is a './' prefix
+      responseType: `stream`,
+    })
```

Comment:
I think it would be good idea to implement some sort of caching mechanism so we don't have to redownload images everytime

sqip plugin implemented this locally - https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sqip/src/extend-node-type.js#L229-L248 - maybe this can be turned into utility that could be moved to main contentful plugin and exported from it?

Reply:
I don't actually download the entire file I just get the metadata from the first bits in the stream and then destroy it

Reply:
Oh, I've missed that, you are of course right. In any case we could cache metadata so we don't do same requests every build I think.

One way to do it is to pass `cache` helper utility here https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-remark/src/extend-node-type.js#L181-L188 and use that here - similar to what sqip plugin is doing - https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sqip/src/generate-sqip.js#L38-L76 (just note that `cache` is not namespaced so, you would have to namespace used keys yourself.

If you are not sure how to use it just let me know - I can jump in and do this part.

Reply:
Hey @pieh That would be nice but I think the `gatsby-transfromer-remark` is not passing the `cache` function to it's plugins.
Does it make sense to add a PR that does that ?